### PR TITLE
Fix logging of plan optimizer

### PR DESCRIPTION
**Pull request by feilong-liu**
## Description
Change the result type of PlanOptimizer, so that it not only returns the new plan node, but also returns whether the optimizer is triggered or not.

## Motivation and Context
Currently, the logging of PlanOptimizer is to compare the old and new plan root with the equal operator. However, this leads to false positives, where any non meaningful changes (for example reorder of outputs etc) will show as optimizer triggered. It also relies on proper equal() override for all data structures in query plans.

In this PR, I change the PlanOptimizer to explicitly return whether an optimizer is triggered or not. I've implemented them for all existing plan optimizers.

## Impact
Fix logging of optimizers, and eliminate false positives.

## Test Plan
Existing unit tests.

## Contributor checklist

- [x] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [x] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [x] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [x] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [x] Adequate tests were added if applicable.
- [x] CI passed.

## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.

```
== RELEASE NOTES ==

General Changes
* Fix logging of optimizer triggering, eliminating false positives where optimizer is logged as triggered but actually not. The PlanOptimizer now return both result plan and whether the plan changed or not.
```

### CODE REVIEW

**Comment by mlyublena** [comment]
Thanks for fixing these!

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java [hunk 1]
```diff
@@ -366,6 +374,7 @@ public PlanNode visitJoin(JoinNode joinNode, RewriteContext<Set<VariableReferenc
                             Optional.empty(),
```
Comment (mlyublena) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java:374]: should we also remove these inlined calls? This should now be properly handled outside in the main optimizer loop

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java [hunk 2]
```diff
@@ -164,10 +164,11 @@ public AddExchanges(Metadata metadata, SqlParser parser, PartitioningProviderMan
     }
 
     @Override
-    public PlanNode optimize(PlanNode plan, Session session, TypeProvider types, VariableAllocator variableAllocator, PlanNodeIdAllocator idAllocator, WarningCollector warningCollector)
+    public PlanOptimizerResult optimize(PlanNode plan, Session session, TypeProvider types, VariableAllocator variableAllocator, PlanNodeIdAllocator idAllocator, WarningCollector warningCollector)
     {
         PlanWithProperties result = new Rewriter(idAllocator, variableAllocator, session, partitioningProviderManager).accept(plan, PreferredProperties.any());
-        return result.getNode();
+        boolean optimizerTriggered = PlanNodeSearcher.searchFrom(result.getNode()).where(node -> node instanceof ExchangeNode && ((ExchangeNode) node).getScope().isRemote()).findFirst().isPresent();
```
Comment (mlyublena) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java:170]: are we sure that this optimizer added the exchange node and it wasn't there before?

**Review by mlyublena — APPROVED** [review]

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java (hunk 1 above)
Comment (feilong-liu) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java:374]: Sure, let me remove this code too.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java (hunk 2 above)
Comment (feilong-liu) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java:170]: Yes, we only have this optimizer to add remote exchange node.

**Review by pranjalssh — APPROVED** [review]
Will all new optimizers will have to implement isOptimizerTriggered method. So all spi optimizers and our internal optimizers api might change. Please add that in release notes

**Pull request merged**